pkg/util/kubernetes/apiserver/apiserver.go

```diff
@@ -225,13 +219,26 @@ func processKubeServices(nodeList *v1.NodeList, podList *v1.PodList, endpointLis
 	for _, node := range nodeList.Items {
 		nodeName := node.Name
 		nodeNameCacheKey := cache.BuildAgentKey(metadataMapperCachePrefix, nodeName)
-		metaBundle, found := cache.Cache.Get(nodeNameCacheKey)
+		freshness := cache.BuildAgentKey(metadataMapperCachePrefix, nodeName, "freshness")
+
+		metaBundle, found := cache.Cache.Get(nodeNameCacheKey)       // We get the old one with the dead pods. if diff reset metabundle and deledte key. Then compute again.
+		freshnessCache, freshnessFound := cache.Cache.Get(freshness) // if expired, freshness not found deal with that
+
 		if !found {
 			metaBundle = newMetadataMapperBundle()
+			cache.Cache.Set(freshness, len(podList.Items), metadataMapExpire)
 		}
+
+		if found && freshnessCache != len(podList.Items) || !freshnessFound {
```

Comment:
If a pod is deleted and a new pod is rescheduled, we may have some issue here no ?

Reply:
@JulienBalestra technically yes, but this is the worst case scenario. 
In this case, the map will have an additional entry for max 40 seconds (we refresh every 20 second and the expiration time is 60 seconds).
The main problem we had with the previous implementation was that the cache was never churned.